src/Bicep.Core/TypeSystem/Providers/ProviderSpecificationSyntaxFactory.cs

```diff
@@ -0,0 +1,82 @@
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex ConfigManagedSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex LegacyBuiltInSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<address>{BicepRegistryAddressPattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex InlinedSpecificationPattern();
+
+    [GeneratedRegex(@"^\S*[:\/](?<name>\S+)$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex RepositoryNamePattern();
+
+    public static IProviderSpecificationSyntax CreateProviderSpecificationSyntax(SyntaxBase syntax)
+     => syntax switch
+     {
+         StringSyntax stringSyntax when stringSyntax.TryGetLiteralValue() is { } value &&
+                                       TryCreateFromStringSyntax(stringSyntax, value) is { } specification
+             => specification,
+
+         IdentifierSyntax identifierSyntax
+             => new ConfigurationManagedProviderSpecificationSyntax(
+                 identifierSyntax.IdentifierName,
+                 IsValid: true,
+                 identifierSyntax.Span),
+
+         _ => new ProviderSpecificationSyntaxTrivia(syntax.Span)
+     };
+
+    private static IProviderSpecificationSyntax? TryCreateFromStringSyntax(StringSyntax stringSyntax, string value)
+    {
+        if (LegacyBuiltInSpecificationPattern().Match(value) is { Success: true } legacyMatch)
+        {
+            var name = legacyMatch.Groups["name"].Value;
+            var version = legacyMatch.Groups["version"].Value;
+            var span = new TextSpan(stringSyntax.Span.Position + 1, name.Length);
+
+            return new LegacyProviderSpecificationSyntax(name, version, IsValid: true, span);
+        }
+        else if (ConfigManagedSpecificationPattern().Match(value) is { Success: true } builtInMatch)
+        {
+            var name = builtInMatch.Groups["name"].Value;
+            var span = new TextSpan(stringSyntax.Span.Position + 1, name.Length);
+
+            return new ConfigurationManagedProviderSpecificationSyntax(name, IsValid: true, span);
+        }
+        else if (InlinedSpecificationPattern().Match(value) is { Success: true } registryMatch)
+        {
+            // NOTE(asilverman): The regex for the registry pattern is intentionally loose since it will be validated by the module resolver.
+            var address = registryMatch.Groups["address"].Value;
+            var version = registryMatch.Groups["version"].Value;
+
+            var span = new TextSpan(stringSyntax.Span.Position + 1, address.Length);
```

Comment:
Rather than doing math here, could we create a getter on `StringSyntax` (e.g. `stringSyntax.InnerSpan`)?

StringSyntax can also represent a multi-line syntax, so `Position+1` will give incorrect results in certain cases:
```bicep
var foo = '''
bar
'''
```